intel-oneapi-compilers: add multiarch include directory

stephenmsachs opened the pull request:
On ubuntu systems `icc` will only add `/usr/include/x86_64-linux-gnu` to the include directories if the `gcc` that `icc` was installed with also reports `x86_64-linux-gnu` as its install target. By default `gcc` installed with Spack uses `config.guess` to determine the install target. For ubuntu systems `config.guess` returns `x86_64-pc-linux-gnu` instead of `x86_64-linux-gnu`.

This patch fixes the include path that case.

Commit:
intel-oneapi-compilers: add multiarch include directory

On ubuntu systems `icc` will only add `/usr/include/x86_64-linux-gnu` to the
include directories if the `gcc` that `icc` was installed with also reports
`x86_64-linux-gnu` as its install target. By default `gcc` installed with Spack
uses `config.guess` to determine the install target. For ubuntu systems
`config.guess` returns `x86_64-pc-linux-gnu` instead of `x86_64-linux-gnu`.

This patch fixes the include path that case.

Commit:
fix format

alalazo commented:
For context, see discussion in https://github.com/spack/spack/pull/40402

alalazo commented:
This LGTM, @rscohn2 can you have a look at this too?

rscohn2 reviewed (dismissed):
Thanks! It looks like a difficult problem.

haampie commented on var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:
```diff
@@ -267,6 +267,15 @@ def extend_config_flags(self):
+        multiarch_dir = gcc("-print-multiarch", output=str).strip()
+        triplet = gcc("-###", error=str).partition("Target: ")[2].split("\n")[0]
+        if multiarch_dir and multiarch_dir != triplet:
+            classic_flags.append("-isystem/usr/include/" + multiarch_dir)
```
`/usr/include` isn't correct on nix, gentoo prefix, etc.

stephenmsachs replied:
Yes, I was wondering about. How do we get that prefix on other systems?

alalazo replied:
It would be better to find the correct prefix, but my understanding is that it wouldn't be a hard error if the path doesn't exist. In any case, you can check if the include dir exists before adding it.

stephenmsachs replied:
Yes, the compilr would not mind if the directory does not exists. I have tried to find out how to get that prefix from `gcc` but I have not found anythign better than `gcc -### | grep configure | grep -- --prefix`. I would not expect that to always be set.

haampie replied:
Yeah, this is going to be a pain :)

I guess nix and gentoo prefix and friends don't use multiarch, exactly cause they have one package per prefix... but who knows they do support a single gcc that can target `gcc -m64 / -m32 / -mx32` on x86... :weary: 

And if they do, then glibc likely lives in a different prefix from gcc, so looking at gcc's prefix isn't useful

stephenmsachs replied:
OK. I have not been able to find anythign in `gcc`, `cpp`, and friends. Also, the reported install prefix is actually wrong for non-system installed ones. How about we only add this if the directory `/usr/include/...` exists?

haampie replied:
What GCC does is this: it has a bunch of config files per platform, they set the variable `MULTILIB_OSDIRNAMES`, this variable ultimately ends up in the output of `gcc -dumpspecs`.

that variable is typically set like:

```
$(call if_multiarch,:x86_64-linux-gnu)
```

where `x86_64-linux-gnu` is a hard-coded triplet, which `gcc -print-multiarch` prints.

Now, the `if_multiarch` make function is defined like this:

```make
if_multiarch = $(if $(wildcard $(shell echo $(BUILD_SYSTEM_HEADER_DIR))/../../usr/lib/*/crti.o),$(1))
```

so... it looks for glibc startup files relative to something, the relevant thing being that `*`, it's in a subdir of lib, not directly in lib.

---

long story short: maybe it's an idea to just let gcc compile `helloworld.c` with `-###`, and figure out the `crti.o` directory from the link line?

On Ubuntu, that is:

```
/usr/lib/gcc/x86_64-linux-gnu/12/../../../x86_64-linux-gnu/crti.o
```

or 

```
/usr/lib/x86_64-linux-gnu/crti.o
```

If the dirname of crti.o isn't lib or lib64, or matches a triplet... use that?

haampie replied:
anyways, everything is awful here :weary:

haampie replied:
Hmm, nevermind, I don't think that works either, cause `if_multiarch` is only defined like that when multiarch is autodetected by GCC; debian / ubuntu just set the flag `--enable-multiarch` explicitly.

So in the end you still have to figure out how gcc determines the `/usr/include` dir, I guess.

haampie replied:
This could work:

```
$ `gcc -print-prog-name=cc1` -imultiarch `gcc -print-multiarch` -v
ignoring nonexistent directory "/usr/local/include/x86_64-linux-gnu"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/12/include-fixed"
ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/12/../../../../x86_64-linux-gnu/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/gcc/x86_64-linux-gnu/12/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
```

So

```
 /usr/lib/gcc/x86_64-linux-gnu/12/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
```
are the relevant include dirs

stephenmsachs replied:
I just saw your last comment. So can we agree on this?
```
`gcc -print-prog-name=cc1` -imultiarch `gcc -print-multiarch` -v /dev/null 2>&1 | grep -v "nonexistent" | grep `gcc -print-multiarch`$
```

haampie reviewed (changes requested):
Don't use `/usr/include`, it doesn't work in general

Commit:
Find gcc default include prefix in case /usr/include does not exist

Commit:
fix format

Commit:
Use cc1 with correct flags to ask for multilib include directories

Use equivalent of:
`gcc -print-prog-name=cc1` -imultiarch `gcc -print-multiarch` -v /dev/null 2>&1 \
      | grep -v "nonexistent" | grep `gcc -print-multiarch`$

Commit:
fix format

hmenke reviewed (approved):
Works for me on Debian 11. Seems to be an okay solution.

haampie committed:
Merge branch 'develop' into icc-multiarch

haampie commented on var/spack/repos/builtin/packages/intel-oneapi-compilers/package.py:
```diff
@@ -518,6 +518,27 @@ def extend_config_flags(self):
+        # correct flags.
+        gcc = Executable(self.compiler.cc)
+        cc1 = Executable(gcc("-print-prog-name=cc1", output=str).strip())
+        multiarch_triplet = gcc("-print-multiarch", output=str).strip()
```
`multiarch_triplet` can be empty string, then `i.endswith("\n")` is always true.

stephenmsachs replied:
I have added a change so that the `incdir` is only going to be created if `multiarch_triplet` is not empty.

Commit:
Attempt to create include dirs only if multiarch triplet exists

alalazo commented:
Closing, since we moved the package repository. Feel free to reopen at / migrate to https://github.com/spack/spack-packages if you want to continue work on this.
